File: python/ctci/urlify.py
import sys

def urlify(s, true_size):
    s = list(s)
    trueindex = true_size-1
    newindex = trueindex + (2*s[:true_size].count(' '))
    while newindex >= 0:
        while trueindex >= 0:
            if s[trueindex] == ' ':
                s[newindex] = '0'
                s[newindex - 1] = '2'
                s[newindex - 2] = '%'
                newindex -= 3
            else:   
                s[newindex] = s[trueindex]
                newindex -= 1
            trueindex -= 1
    return "".join(s)

if __name__ == '__main__':

    for line in sys.stdin.readlines():
        line = line.strip()
        print("The url of {} is {}".format(line, urlify(line.ljust(30, ' '), len(line))))
        print()

File: cpp/ctci/urlify.cpp
#include <iostream>
#include <string>
#include <cstring>
using namespace std;
    
char* urlify(char* str, int true_size) {
    true_size--;
    int end_index = true_size;
    for (int i = 0; i <= true_size; i++) {
        if (str[i] == ' ') end_index += 2;
    }
    for (int i = 0; i <= true_size; i++) {
        if(str[true_size - i] == ' ') {
            str[end_index] = '0';
            str[end_index - 1] = '2';
            str[end_index - 2] = '%';
            end_index -= 3;
        } else {
            str[end_index] = str[true_size - i];
            end_index--;
        }
    }
    return str;
}

int main() {
    string s;
    char str[30];
    while (getline(cin, s)) {
        strcpy(str, s.c_str());
        cout << "The url of " << s << " is " << urlify(str, s.size()) << endl;
    }
}
